

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Seq/ICTV/MSL and VMR &mdash; mtbp3  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
      <link rel="stylesheet" type="text/css" href="_static/graphviz.css?v=fd3f3429" />
      <link rel="stylesheet" type="text/css" href="_static/css/custom.css?v=ce7ddcad" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Changelog" href="changelog.html" />
    <link rel="prev" title="&lt;no title&gt;" href="statlab_corr_spearman_rho.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            mtbp3
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="example_cdtg.html">U/Graph/Categorical Plots Using catPlotter</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_files.html">U/System/List Folder Contents</a></li>
<li class="toctree-l1"><a class="reference internal" href="util_jsonutil.html">U/Data/JSON tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_clinical.html">H/Clinical/Grouped Summary</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_ectd.html">H/ICH/eCTD/US/3.2/Heading</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_ectd2.html">H/ICH/eCTD/US/4.0/Heading</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_emt.html">H/ICH/MedDRA/Basics</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_emt2.html">H/ICH/MedDRA/SMQs by ICH</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_emt3.html">H/ICH/MedDRA/FMQs by FDA</a></li>
<li class="toctree-l1"><a class="reference internal" href="example_emt4.html">H/ICH/MedDRA/(MLGs by CIOMS)</a></li>
<li class="toctree-l1"><a class="reference internal" href="std_cdisc.html">S/CDISC/Library Resource</a></li>
<li class="toctree-l1"><a class="reference internal" href="std_hl7fhir_us_pqcmc.html">S/HL7/FHIR-R5/US/PQ-CMC</a></li>
<li class="toctree-l1"><a class="reference internal" href="std_iso_pdf.html">S/ISO/PDF</a></li>
<li class="toctree-l1"><a class="reference internal" href="std_iso_idmp.html">S/ISO/IDMP</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Seq/ICTV/MSL and VMR</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#an-example">An Example</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-to">How to</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#search-within-msl">Search within MSL</a></li>
<li class="toctree-l3"><a class="reference internal" href="#download-updated-msl">Download Updated MSL</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#subtypes-within-species">Subtypes within Species</a></li>
<li class="toctree-l2"><a class="reference internal" href="#species-exemplar">Species Exemplar</a></li>
<li class="toctree-l2"><a class="reference internal" href="#other-resource">Other Resource</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference">Reference</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="conduct.html">Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference internal" href="autoapi/index.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">mtbp3</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Seq/ICTV/MSL and VMR</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/seq_ictv_taxonomy.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="seq-ictv-msl-and-vmr">
<h1>Seq/ICTV/MSL and VMR<a class="headerlink" href="#seq-ictv-msl-and-vmr" title="Link to this heading"></a></h1>
<p>International Committee on Taxonomy of Viruses (ICTV) <a class="footnote-reference brackets" href="#id16" id="id1" role="doc-noteref"><span class="fn-bracket">[</span>1<span class="fn-bracket">]</span></a> maintains a database <a class="footnote-reference brackets" href="#id17" id="id2" role="doc-noteref"><span class="fn-bracket">[</span>2<span class="fn-bracket">]</span></a> of virus taxonomic classification <a class="footnote-reference brackets" href="#id18" id="id3" role="doc-noteref"><span class="fn-bracket">[</span>3<span class="fn-bracket">]</span></a>.
The Master Species Lists (MSL) <a class="footnote-reference brackets" href="#id19" id="id4" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a> provides “the official, current virus taxonomy approved by the ICTV”.
The Virus Metadata Resource (VMR) <a class="footnote-reference brackets" href="#id20" id="id5" role="doc-noteref"><span class="fn-bracket">[</span>5<span class="fn-bracket">]</span></a> provides “exemplars and additional isolates for each virus species”.</p>
<section id="an-example">
<h2>An Example<a class="headerlink" href="#an-example" title="Link to this heading"></a></h2>
<p>ICTV provides MSL downloads in .xlsx <a class="footnote-reference brackets" href="#id19" id="id6" role="doc-noteref"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></a>, web version of interactive search <a class="footnote-reference brackets" href="#id21" id="id7" role="doc-noteref"><span class="fn-bracket">[</span>6<span class="fn-bracket">]</span></a>, and visual taxonomy browser <a class="footnote-reference brackets" href="#id22" id="id8" role="doc-noteref"><span class="fn-bracket">[</span>7<span class="fn-bracket">]</span></a>.
For example, if we are interested in studying the virus causing feline acquired immunodeficiency syndrome <a class="footnote-reference brackets" href="#id23" id="id9" role="doc-noteref"><span class="fn-bracket">[</span>8<span class="fn-bracket">]</span></a>, we can find the following taxonomy information in the ICTV MSL:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Column name</p></th>
<th class="head"><p>value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Realm</p></td>
<td><p>Riboviria</p></td>
</tr>
<tr class="row-odd"><td><p>Subrealm</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>Kingdom</p></td>
<td><p>Pararnavirae</p></td>
</tr>
<tr class="row-odd"><td><p>Subkingdom</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>Phylum</p></td>
<td><p>Artverviricota</p></td>
</tr>
<tr class="row-odd"><td><p>Subphylum</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>Class</p></td>
<td><p>Revtraviricetes</p></td>
</tr>
<tr class="row-odd"><td><p>Subclass</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>Order</p></td>
<td><p>Ortervirales</p></td>
</tr>
<tr class="row-odd"><td><p>Suborder</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>Family</p></td>
<td><p>Retroviridae</p></td>
</tr>
<tr class="row-odd"><td><p>Subfamily</p></td>
<td><p>Orthoretrovirinae</p></td>
</tr>
<tr class="row-even"><td><p>Genus</p></td>
<td><p>Lentivirus</p></td>
</tr>
<tr class="row-odd"><td><p>Subgenus</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>Species</p></td>
<td><p>Lentivirus felimdef</p></td>
</tr>
<tr class="row-odd"><td><p>Genome</p></td>
<td><p>ssRNA-RT</p></td>
</tr>
<tr class="row-even"><td><p>Last Change</p></td>
<td><p>Renamed,</p></td>
</tr>
<tr class="row-odd"><td><p>MSL of Last Change</p></td>
<td><p>39</p></td>
</tr>
<tr class="row-even"><td><p>Proposal for Last Change</p></td>
<td><p>2023.009D.Retroviridae_68rensp.zip</p></td>
</tr>
<tr class="row-odd"><td><p>Taxon History URL</p></td>
<td><p>ictv.global=202305029</p></td>
</tr>
</tbody>
</table>
<p>From the taxon history, we can see that the species name was <em>Feline immunodeficiency virus</em> (FIV) in the previous releases, and changed to <em>Lentivirus felimdef</em> in the 2023 release <a class="footnote-reference brackets" href="#id24" id="id10" role="doc-noteref"><span class="fn-bracket">[</span>9<span class="fn-bracket">]</span></a>.
The reasons of changes proposed are included in the <em>Proposal for Last Change</em> link.
Additional information about FIV, for example, about the genus <em>Lentivirus</em>, can be found in ICTV report <a class="footnote-reference brackets" href="#id25" id="id11" role="doc-noteref"><span class="fn-bracket">[</span>10<span class="fn-bracket">]</span></a>.</p>
</section>
<section id="how-to">
<h2>How to<a class="headerlink" href="#how-to" title="Link to this heading"></a></h2>
<p>The module <code class="docutils literal notranslate"><span class="pre">mtbp3.seq.ictvmsiview</span></code> accept a csv file exported from the MSL tab of MSL download file.
The csv file path can be assigned using the option <code class="docutils literal notranslate"><span class="pre">msl_file_path</span></code>.
For demonstration, a small subset of MSL is shipped with this module.
When using the option <code class="docutils literal notranslate"><span class="pre">msl_file_path</span> <span class="pre">=</span> <span class="pre">&quot;&quot;</span></code>, the example file will be loaded.</p>
<p>To load a MSL csv file:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">mtbp3.seq</span> <span class="kn">import</span> <span class="n">ictvmslview</span>
<span class="n">msl</span> <span class="o">=</span> <span class="n">ictvmslview</span><span class="o">.</span><span class="n">ictvmsl</span><span class="p">(</span><span class="n">msl_file_path</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File supp_seq/ICTV_MSL39v4_example.csv has been loaded
Column names: [&#39;Sort&#39;, &#39;Realm&#39;, &#39;Subrealm&#39;, &#39;Kingdom&#39;, &#39;Subkingdom&#39;, &#39;Phylum&#39;, &#39;Subphylum&#39;, &#39;Class&#39;, &#39;Subclass&#39;, &#39;Order&#39;, &#39;Suborder&#39;, &#39;Family&#39;, &#39;Subfamily&#39;, &#39;Genus&#39;, &#39;Subgenus&#39;, &#39;Species&#39;, &#39;Genome&#39;, &#39;Last Change&#39;, &#39;MSL of Last Change&#39;, &#39;Proposal for Last Change &#39;, &#39;Taxon History URL&#39;]
Total number of rows: 65
</pre></div>
</div>
</div>
</div>
<p>To view the first row of loaded file:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">msl</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sort                                                      13664
Realm                                                 Riboviria
Subrealm                                                    NaN
Kingdom                                            Pararnavirae
Subkingdom                                                  NaN
Phylum                                           Artverviricota
Subphylum                                                   NaN
Class                                           Revtraviricetes
Subclass                                                    NaN
Order                                              Ortervirales
Suborder                                                    NaN
Family                                             Retroviridae
Subfamily                                     Orthoretrovirinae
Genus                                           Alpharetrovirus
Subgenus                                                    NaN
Species                           Alpharetrovirus avicarmilhil2
Genome                                                 ssRNA-RT
Last Change                                            Renamed,
MSL of Last Change                                           39
Proposal for Last Change     2023.009D.Retroviridae_68rensp.zip
Taxon History URL                         ictv.global=202304982
Name: 0, dtype: object
</pre></div>
</div>
</div>
</div>
<p>To view the unique values of ‘Realm’ in the file and number of rows under each realm:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">msl</span><span class="o">.</span><span class="n">msl</span><span class="p">[</span><span class="s1">&#39;Realm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Realm
Riboviria    65
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>The counts above are from the example file.
The full MSL 39v4 includes the following realms:</p>
<table class="docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Realm</p></th>
<th class="head"><p>Count of Species</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Adnaviria</p></td>
<td><p>32</p></td>
</tr>
<tr class="row-odd"><td><p>Duplodnaviria</p></td>
<td><p>4973</p></td>
</tr>
<tr class="row-even"><td><p>Monodnaviria</p></td>
<td><p>2028</p></td>
</tr>
<tr class="row-odd"><td><p>Riboviria</p></td>
<td><p>6712</p></td>
</tr>
<tr class="row-even"><td><p>Ribozyviria</p></td>
<td><p>21</p></td>
</tr>
<tr class="row-odd"><td><p>Varidnaviria</p></td>
<td><p>288</p></td>
</tr>
</tbody>
</table>
<p>Please note that the column names will be used as <code class="docutils literal notranslate"><span class="pre">search_rank</span></code> in this module and those are case sensitive.
The <code class="docutils literal notranslate"><span class="pre">search_strings</span></code> is not case sensitive, as shown below.</p>
<section id="search-within-msl">
<h3>Search within MSL<a class="headerlink" href="#search-within-msl" title="Link to this heading"></a></h3>
<p>To search for rows with “lentivirus fel” included in a species name:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="s2">&quot;lentivirus fel&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">transpose</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Sort                                                      13704
Realm                                                 Riboviria
Subrealm                                                    NaN
Kingdom                                            Pararnavirae
Subkingdom                                                  NaN
Phylum                                           Artverviricota
Subphylum                                                   NaN
Class                                           Revtraviricetes
Subclass                                                    NaN
Order                                              Ortervirales
Suborder                                                    NaN
Family                                             Retroviridae
Subfamily                                     Orthoretrovirinae
Genus                                                Lentivirus
Subgenus                                                    NaN
Species                            <span class=" -Color -Color-Red">Lentivirus fel</span>imdef
Genome                                                 ssRNA-RT
Last Change                                            Renamed,
MSL of Last Change                                           39
Proposal for Last Change     2023.009D.Retroviridae_68rensp.zip
Taxon History URL                         ictv.global=202305029
Name: 40, dtype: object
</pre></div>
</div>
</div>
</div>
<p>To search for rows with “lentivirus” included in species:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="s2">&quot;lentivirus&quot;</span><span class="p">,</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Species&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">narrow</span><span class="o">=</span><span class="kc">True</span><span class="p">)[[</span><span class="s1">&#39;Genus&#39;</span><span class="p">,</span> <span class="s1">&#39;Species&#39;</span><span class="p">,</span><span class="s1">&#39;Genome&#39;</span><span class="p">]])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>         Genus                        Species    Genome
36  Lentivirus   <span class=" -Color -Color-Red">Lentivirus</span> bovimdef  ssRNA-RT
37  Lentivirus     <span class=" -Color -Color-Red">Lentivirus</span> bovjem  ssRNA-RT
38  Lentivirus  <span class=" -Color -Color-Red">Lentivirus</span> capartenc  ssRNA-RT
39  Lentivirus  <span class=" -Color -Color-Red">Lentivirus</span> equinfane  ssRNA-RT
40  Lentivirus   <span class=" -Color -Color-Red">Lentivirus</span> felimdef  ssRNA-RT
41  Lentivirus  <span class=" -Color -Color-Red">Lentivirus</span> humimdef1  ssRNA-RT
42  Lentivirus  <span class=" -Color -Color-Red">Lentivirus</span> humimdef2  ssRNA-RT
43  Lentivirus  <span class=" -Color -Color-Red">Lentivirus</span> ovivismae  ssRNA-RT
44  Lentivirus        <span class=" -Color -Color-Red">Lentivirus</span> pum  ssRNA-RT
45  Lentivirus   <span class=" -Color -Color-Red">Lentivirus</span> simimdef  ssRNA-RT
</pre></div>
</div>
</div>
</div>
<p>To search for rows with “lentivirus” as Genus using <code class="docutils literal notranslate"><span class="pre">exact=True</span></code> option (partly exact, the searching is still not case sensitive):</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="s2">&quot;lentivirus&quot;</span><span class="p">,</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Genus&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">,</span> <span class="n">exact</span><span class="o">=</span><span class="kc">True</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    [Realm] Riboviria:
    └── [Kingdom] Pararnavirae:
        └── [Phylum] Artverviricota:
            └── [Class] Revtraviricetes:
                └── [Order] Ortervirales:
                    └── [Family] Retroviridae; [Subfamily] Orthoretrovirinae:
                        └── [Genus] <span class=" -Color -Color-Red">Lentivirus</span>:
                            ├── [Species] Lentivirus bovimdef
                            ├── [Species] Lentivirus bovjem
                            ├── [Species] Lentivirus capartenc
                            ├── [Species] Lentivirus equinfane
                            ├── [Species] Lentivirus felimdef
                            ├── [Species] Lentivirus humimdef1
                            ├── [Species] Lentivirus humimdef2
                            ├── [Species] Lentivirus ovivismae
                            ├── [Species] Lentivirus pum
                            └── [Species] Lentivirus simimdef
</pre></div>
</div>
</div>
</div>
<p>There are more than one virus species in the Retroviridae family that can infect feline.
To search for virus within the Retroviridae famiy that can infect feline:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="s2">&quot; fel&quot;</span><span class="p">,</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Species&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">,</span> <span class="n">search_within_subset</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Family&quot;</span><span class="p">:</span><span class="s2">&quot;Retroviridae&quot;</span><span class="p">})))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    [Realm] Riboviria:
    └── [Kingdom] Pararnavirae:
        └── [Phylum] Artverviricota:
            └── [Class] Revtraviricetes:
                └── [Order] Ortervirales:
                    ├── [Family] Retroviridae; [Subfamily] Orthoretrovirinae:
                    │   ├── [Genus] Gammaretrovirus:
                    │   │   └── [Species] Gammaretrovirus<span class=" -Color -Color-Red"> fel</span>leu
                    │   └── [Genus] Lentivirus:
                    │       └── [Species] Lentivirus<span class=" -Color -Color-Red"> fel</span>imdef
                    └── [Family] Retroviridae; [Subfamily] Spumaretrovirinae:
                        └── [Genus] Felispumavirus:
                            └── [Species] Felispumavirus<span class=" -Color -Color-Red"> fel</span>cat
</pre></div>
</div>
</div>
</div>
<p>To search for two known virus species:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gammaretrovirus felleu&quot;</span><span class="p">,</span> <span class="s2">&quot;Lentivirus felimdef&quot;</span><span class="p">],</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Species&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    [Realm] Riboviria:
    └── [Kingdom] Pararnavirae:
        └── [Phylum] Artverviricota:
            └── [Class] Revtraviricetes:
                └── [Order] Ortervirales:
                    └── [Family] Retroviridae; [Subfamily] Orthoretrovirinae:
                        ├── [Genus] Gammaretrovirus:
                        │   └── [Species] <span class=" -Color -Color-Red">Gammaretrovirus felleu</span>
                        └── [Genus] Lentivirus:
                            └── [Species] <span class=" -Color -Color-Red">Lentivirus felimdef</span>
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>There are two more types of tree available, including a tree with nonempty ranks:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gammaretrovirus felleu&quot;</span><span class="p">,</span> <span class="s2">&quot;Lentivirus felimdef&quot;</span><span class="p">],</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Species&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="s2">&quot;drop&quot;</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    [Realm] Riboviria:
    └── [Kingdom] Pararnavirae:
        └── [Phylum] Artverviricota:
            └── [Class] Revtraviricetes:
                └── [Order] Ortervirales:
                    └── [Family] Retroviridae:
                        └── [Subfamily] Orthoretrovirinae:
                            ├── [Genus] Gammaretrovirus:
                            │   └── [Species] <span class=" -Color -Color-Red">Gammaretrovirus felleu</span>
                            └── [Genus] Lentivirus:
                                └── [Species] <span class=" -Color -Color-Red">Lentivirus felimdef</span>
</pre></div>
</div>
</div>
</div>
<p>and a tree with full 15 ranks:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">msl</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gammaretrovirus felleu&quot;</span><span class="p">,</span> <span class="s2">&quot;Lentivirus felimdef&quot;</span><span class="p">],</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Species&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="s2">&quot;full&quot;</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>    [Realm] Riboviria:
    └── [Subrealm] NA:
        └── [Kingdom] Pararnavirae:
            └── [Subkingdom] NA:
                └── [Phylum] Artverviricota:
                    └── [Subphylum] NA:
                        └── [Class] Revtraviricetes:
                            └── [Subclass] NA:
                                └── [Order] Ortervirales:
                                    └── [Suborder] NA:
                                        └── [Family] Retroviridae:
                                            └── [Subfamily] Orthoretrovirinae:
                                                ├── [Genus] Gammaretrovirus:
                                                │   └── [Subgenus] NA:
                                                │       └── [Species] <span class=" -Color -Color-Red">Gammaretrovirus felleu</span>
                                                └── [Genus] Lentivirus:
                                                    └── [Subgenus] NA:
                                                        └── [Species] <span class=" -Color -Color-Red">Lentivirus felimdef</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="download-updated-msl">
<h3>Download Updated MSL<a class="headerlink" href="#download-updated-msl" title="Link to this heading"></a></h3>
<p>To download the current release, use <code class="docutils literal notranslate"><span class="pre">msl.update_msl(version=&quot;current&quot;)</span></code>.
That will return output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">File</span> <span class="n">of</span> <span class="n">version</span> <span class="n">current</span> <span class="n">has</span> <span class="n">been</span> <span class="n">loaded</span>
<span class="n">Column</span> <span class="n">names</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Sort&#39;</span><span class="p">,</span> <span class="s1">&#39;Realm&#39;</span><span class="p">,</span> <span class="s1">&#39;Subrealm&#39;</span><span class="p">,</span> <span class="s1">&#39;Kingdom&#39;</span><span class="p">,</span> <span class="s1">&#39;Subkingdom&#39;</span><span class="p">,</span> <span class="s1">&#39;Phylum&#39;</span><span class="p">,</span> <span class="s1">&#39;Subphylum&#39;</span><span class="p">,</span> <span class="s1">&#39;Class&#39;</span><span class="p">,</span> <span class="s1">&#39;Subclass&#39;</span><span class="p">,</span> <span class="s1">&#39;Order&#39;</span><span class="p">,</span> <span class="s1">&#39;Suborder&#39;</span><span class="p">,</span> <span class="s1">&#39;Family&#39;</span><span class="p">,</span> <span class="s1">&#39;Subfamily&#39;</span><span class="p">,</span> <span class="s1">&#39;Genus&#39;</span><span class="p">,</span> <span class="s1">&#39;Subgenus&#39;</span><span class="p">,</span> <span class="s1">&#39;Species&#39;</span><span class="p">,</span> <span class="s1">&#39;Genome&#39;</span><span class="p">,</span> <span class="s1">&#39;Last Change&#39;</span><span class="p">,</span> <span class="s1">&#39;MSL of Last Change&#39;</span><span class="p">,</span> <span class="s1">&#39;Proposal for Last Change &#39;</span><span class="p">,</span> <span class="s1">&#39;Taxon History URL&#39;</span><span class="p">]</span>
<span class="n">Total</span> <span class="n">number</span> <span class="n">of</span> <span class="n">rows</span><span class="p">:</span> <span class="mi">14690</span>
</pre></div>
</div>
<p>To see versions available to download:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">msl</span><span class="o">.</span><span class="n">update_msl</span><span class="p">(</span><span class="n">version</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Version &#39;&#39; is not supported. Supported versions are: current, 39.v4, 39.v3, 39.v2, 39.v1, 38.v3, 38.v2, 38.v1, 2021.v3, 2021.v2, 2021.v1, 2020, 2019, 2018b.v2, 2018a, 2017, 2016.v1.3, 2015, 2014.v4, 2013.v2, 2012.v4, 2011.v2, 2009.v10, 2008, 2005.v1
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="subtypes-within-species">
<h2>Subtypes within Species<a class="headerlink" href="#subtypes-within-species" title="Link to this heading"></a></h2>
<p>ICTV MSL focuses on “taxa at and above the species rank” <a class="footnote-reference brackets" href="#id26" id="id12" role="doc-noteref"><span class="fn-bracket">[</span>11<span class="fn-bracket">]</span></a>.
Classification below the species are often related to species specific characteristics <a class="footnote-reference brackets" href="#id27" id="id13" role="doc-noteref"><span class="fn-bracket">[</span>12<span class="fn-bracket">]</span></a>.</p>
</section>
<section id="species-exemplar">
<h2>Species Exemplar<a class="headerlink" href="#species-exemplar" title="Link to this heading"></a></h2>
<p>ICTV VMR provides one exemplar (and may be more additional isolates) for each species in VMR, with Genebak accession number and direct link to NCBI database.</p>
<p>To load a VRM table extracted from VRM download file:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">mtbp3.seq</span> <span class="kn">import</span> <span class="n">ictvvmrview</span>
<span class="n">vmr</span> <span class="o">=</span> <span class="n">ictvvmrview</span><span class="o">.</span><span class="n">ictvvmr</span><span class="p">(</span><span class="n">vmr_file_path</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>File supp_seq/ICTV_VMR_MSL39v4_example.csv has been loaded
Column names: [&#39;Isolate ID&#39;, &#39;Species Sort&#39;, &#39;Isolate Sort&#39;, &#39;Realm&#39;, &#39;Subrealm&#39;, &#39;Kingdom&#39;, &#39;Subkingdom&#39;, &#39;Phylum&#39;, &#39;Subphylum&#39;, &#39;Class&#39;, &#39;Subclass&#39;, &#39;Order&#39;, &#39;Suborder&#39;, &#39;Family&#39;, &#39;Subfamily&#39;, &#39;Genus&#39;, &#39;Subgenus&#39;, &#39;Species&#39;, &#39;Exemplar or additional isolate&#39;, &#39;Virus name(s)&#39;, &#39;Virus name abbreviation(s)&#39;, &#39;Virus isolate designation&#39;, &#39;Virus GENBANK accession&#39;, &#39;Genome coverage&#39;, &#39;Genome&#39;, &#39;Host source&#39;, &#39;Accessions Link&#39;]
Total number of rows: 65
</pre></div>
</div>
</div>
</div>
<p>When the path is empty, an example file will be used for illustration.
Similarily, use <code class="docutils literal notranslate"><span class="pre">vmr.update_vmr(version=&quot;current&quot;)</span></code>. to download the current version from ICTV.</p>
<p>To search for genus “lentivirus”:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vmr</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="s2">&quot;lentivirus&quot;</span><span class="p">,</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Genus&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">,</span> <span class="n">exact</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="s2">&quot;drop&quot;</span><span class="p">)))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span><span class="gt">---------------------------------------------------------------------------</span>
<span class="ne">KeyError</span><span class="g g-Whitespace">                                  </span>Traceback (most recent call last)
<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/indexes/base.py:3805,</span> in <span class="ni">Index.get_loc</span><span class="nt">(self, key)</span>
<span class="g g-Whitespace">   </span><span class="mi">3804</span> <span class="k">try</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">3805</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_engine</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">casted_key</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3806</span> <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">err</span><span class="p">:</span>

<span class="nn">File index.pyx:167,</span> in <span class="ni">pandas._libs.index.IndexEngine.get_loc</span><span class="nt">()</span>

<span class="nn">File index.pyx:196,</span> in <span class="ni">pandas._libs.index.IndexEngine.get_loc</span><span class="nt">()</span>

<span class="nn">File pandas/_libs/hashtable_class_helper.pxi:2606,</span> in <span class="ni">pandas._libs.hashtable.Int64HashTable.get_item</span><span class="nt">()</span>

<span class="nn">File pandas/_libs/hashtable_class_helper.pxi:2630,</span> in <span class="ni">pandas._libs.hashtable.Int64HashTable.get_item</span><span class="nt">()</span>

<span class="ne">KeyError</span>: 19

<span class="n">The</span> <span class="n">above</span> <span class="n">exception</span> <span class="n">was</span> <span class="n">the</span> <span class="n">direct</span> <span class="n">cause</span> <span class="n">of</span> <span class="n">the</span> <span class="n">following</span> <span class="n">exception</span><span class="p">:</span>

<span class="ne">KeyError</span><span class="g g-Whitespace">                                  </span>Traceback (most recent call last)
<span class="n">Cell</span> <span class="n">In</span><span class="p">[</span><span class="mi">13</span><span class="p">],</span> <span class="n">line</span> <span class="mi">1</span>
<span class="ne">----&gt; </span><span class="mi">1</span> <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vmr</span><span class="o">.</span><span class="n">find_rows_given_str</span><span class="p">(</span><span class="n">search_strings</span><span class="o">=</span><span class="s2">&quot;lentivirus&quot;</span><span class="p">,</span> <span class="n">search_rank</span><span class="o">=</span><span class="s2">&quot;Genus&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">outfmt</span><span class="o">=</span><span class="s1">&#39;tree&#39;</span><span class="p">,</span> <span class="n">exact</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="s2">&quot;drop&quot;</span><span class="p">)))</span>

<span class="nn">File ~/work/mtbp3/mtbp3/mtbp3/seq/ictvvmrview.py:218,</span> in <span class="ni">ictvvmr.find_rows_given_str</span><span class="nt">(self, search_strings, search_rank, color, narrow, outfmt, exact, search_within_subset, tree_style)</span>
<span class="g g-Whitespace">    </span><span class="mi">216</span>     <span class="k">return</span> <span class="n">filtered_df</span>
<span class="g g-Whitespace">    </span><span class="mi">217</span> <span class="k">elif</span> <span class="n">outfmt</span> <span class="o">==</span> <span class="s2">&quot;tree&quot;</span><span class="p">:</span>
<span class="ne">--&gt; </span><span class="mi">218</span>     <span class="n">filtered_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_narrow</span><span class="p">(</span><span class="n">filtered_df</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">tree_style</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">219</span>     <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">220</span>     <span class="n">filtered_df</span> <span class="o">=</span> <span class="n">filtered_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">3</span><span class="p">:]</span>

<span class="nn">File ~/work/mtbp3/mtbp3/mtbp3/seq/ictvvmrview.py:115,</span> in <span class="ni">ictvvmr.make_narrow</span><span class="nt">(vmr, method)</span>
<span class="g g-Whitespace">    </span><span class="mi">113</span>     <span class="n">index1</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="s1">&#39;Virus name(s)&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">114</span>     <span class="n">index2</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="s1">&#39;Virus isolate designation&#39;</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">115</span>     <span class="n">vmr</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">index1</span><span class="p">]</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="n">index1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;\/&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">index1</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="n">index1</span><span class="p">])</span>
<span class="g g-Whitespace">    </span><span class="mi">116</span>     <span class="n">vmr</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">index2</span><span class="p">]</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="n">index2</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;\/&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">index2</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="n">index2</span><span class="p">])</span>
<span class="g g-Whitespace">    </span><span class="mi">117</span> <span class="k">if</span> <span class="s1">&#39;Virus name(s)&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Virus isolate designation&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Virus GENBANK accession&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Exemplar or additional isolate&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/frame.py:10374,</span> in <span class="ni">DataFrame.apply</span><span class="nt">(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)</span>
<span class="g g-Whitespace">  </span><span class="mi">10360</span> <span class="kn">from</span> <span class="nn">pandas.core.apply</span> <span class="kn">import</span> <span class="n">frame_apply</span>
<span class="g g-Whitespace">  </span><span class="mi">10362</span> <span class="n">op</span> <span class="o">=</span> <span class="n">frame_apply</span><span class="p">(</span>
<span class="g g-Whitespace">  </span><span class="mi">10363</span>     <span class="bp">self</span><span class="p">,</span>
<span class="g g-Whitespace">  </span><span class="mi">10364</span>     <span class="n">func</span><span class="o">=</span><span class="n">func</span><span class="p">,</span>
   <span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="g g-Whitespace">  </span><span class="mi">10372</span>     <span class="n">kwargs</span><span class="o">=</span><span class="n">kwargs</span><span class="p">,</span>
<span class="g g-Whitespace">  </span><span class="mi">10373</span> <span class="p">)</span>
<span class="ne">&gt; </span><span class="mi">10374</span> <span class="k">return</span> <span class="n">op</span><span class="o">.</span><span class="n">apply</span><span class="p">()</span><span class="o">.</span><span class="n">__finalize__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;apply&quot;</span><span class="p">)</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/apply.py:916,</span> in <span class="ni">FrameApply.apply</span><span class="nt">(self)</span>
<span class="g g-Whitespace">    </span><span class="mi">913</span> <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">raw</span><span class="p">:</span>
<span class="g g-Whitespace">    </span><span class="mi">914</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_raw</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">engine</span><span class="p">,</span> <span class="n">engine_kwargs</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">engine_kwargs</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">916</span> <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_standard</span><span class="p">()</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/apply.py:1063,</span> in <span class="ni">FrameApply.apply_standard</span><span class="nt">(self)</span>
<span class="g g-Whitespace">   </span><span class="mi">1061</span> <span class="k">def</span> <span class="nf">apply_standard</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1062</span>     <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">engine</span> <span class="o">==</span> <span class="s2">&quot;python&quot;</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1063</span>         <span class="n">results</span><span class="p">,</span> <span class="n">res_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_series_generator</span><span class="p">()</span>
<span class="g g-Whitespace">   </span><span class="mi">1064</span>     <span class="k">else</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">1065</span>         <span class="n">results</span><span class="p">,</span> <span class="n">res_index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">apply_series_numba</span><span class="p">()</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/apply.py:1081,</span> in <span class="ni">FrameApply.apply_series_generator</span><span class="nt">(self)</span>
<span class="g g-Whitespace">   </span><span class="mi">1078</span> <span class="k">with</span> <span class="n">option_context</span><span class="p">(</span><span class="s2">&quot;mode.chained_assignment&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1079</span>     <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">series_gen</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1080</span>         <span class="c1"># ignore SettingWithCopy here in case the user mutates</span>
<span class="ne">-&gt; </span><span class="mi">1081</span>         <span class="n">results</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">func</span><span class="p">(</span><span class="n">v</span><span class="p">,</span> <span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">kwargs</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1082</span>         <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ABCSeries</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1083</span>             <span class="c1"># If we have a view on v, we need to make a copy because</span>
<span class="g g-Whitespace">   </span><span class="mi">1084</span>             <span class="c1">#  series_generator will swap out the underlying data</span>
<span class="g g-Whitespace">   </span><span class="mi">1085</span>             <span class="n">results</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="nn">File ~/work/mtbp3/mtbp3/mtbp3/seq/ictvvmrview.py:115,</span> in <span class="ni">ictvvmr.make_narrow.&lt;locals&gt;.&lt;lambda&gt;</span><span class="nt">(row)</span>
<span class="g g-Whitespace">    </span><span class="mi">113</span>     <span class="n">index1</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="s1">&#39;Virus name(s)&#39;</span><span class="p">)</span>
<span class="g g-Whitespace">    </span><span class="mi">114</span>     <span class="n">index2</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="s1">&#39;Virus isolate designation&#39;</span><span class="p">)</span>
<span class="ne">--&gt; </span><span class="mi">115</span>     <span class="n">vmr</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">index1</span><span class="p">]</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="n">index1</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;\/&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">index1</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="n">index1</span><span class="p">])</span>
<span class="g g-Whitespace">    </span><span class="mi">116</span>     <span class="n">vmr</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="n">index2</span><span class="p">]</span> <span class="o">=</span> <span class="n">vmr</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="n">index2</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;\/&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">index2</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">row</span><span class="p">[</span><span class="n">index2</span><span class="p">])</span>
<span class="g g-Whitespace">    </span><span class="mi">117</span> <span class="k">if</span> <span class="s1">&#39;Virus name(s)&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Virus isolate designation&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Virus GENBANK accession&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Exemplar or additional isolate&#39;</span> <span class="ow">in</span> <span class="n">vmr</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/series.py:1121,</span> in <span class="ni">Series.__getitem__</span><span class="nt">(self, key)</span>
<span class="g g-Whitespace">   </span><span class="mi">1118</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
<span class="g g-Whitespace">   </span><span class="mi">1120</span> <span class="k">elif</span> <span class="n">key_is_scalar</span><span class="p">:</span>
<span class="ne">-&gt; </span><span class="mi">1121</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_value</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1123</span> <span class="c1"># Convert generator to list before going through hashable part</span>
<span class="g g-Whitespace">   </span><span class="mi">1124</span> <span class="c1"># (We will iterate through the generator there to check for slices)</span>
<span class="g g-Whitespace">   </span><span class="mi">1125</span> <span class="k">if</span> <span class="n">is_iterator</span><span class="p">(</span><span class="n">key</span><span class="p">):</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/series.py:1237,</span> in <span class="ni">Series._get_value</span><span class="nt">(self, label, takeable)</span>
<span class="g g-Whitespace">   </span><span class="mi">1234</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="p">[</span><span class="n">label</span><span class="p">]</span>
<span class="g g-Whitespace">   </span><span class="mi">1236</span> <span class="c1"># Similar to Index.get_value, but we do not fall back to positional</span>
<span class="ne">-&gt; </span><span class="mi">1237</span> <span class="n">loc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_loc</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">1239</span> <span class="k">if</span> <span class="n">is_integer</span><span class="p">(</span><span class="n">loc</span><span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">1240</span>     <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_values</span><span class="p">[</span><span class="n">loc</span><span class="p">]</span>

<span class="nn">File ~/.cache/pypoetry/virtualenvs/mtbp3-STzYEG9e-py3.10/lib/python3.10/site-packages/pandas/core/indexes/base.py:3812,</span> in <span class="ni">Index.get_loc</span><span class="nt">(self, key)</span>
<span class="g g-Whitespace">   </span><span class="mi">3807</span>     <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">casted_key</span><span class="p">,</span> <span class="nb">slice</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
<span class="g g-Whitespace">   </span><span class="mi">3808</span>         <span class="nb">isinstance</span><span class="p">(</span><span class="n">casted_key</span><span class="p">,</span> <span class="n">abc</span><span class="o">.</span><span class="n">Iterable</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3809</span>         <span class="ow">and</span> <span class="nb">any</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">slice</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">casted_key</span><span class="p">)</span>
<span class="g g-Whitespace">   </span><span class="mi">3810</span>     <span class="p">):</span>
<span class="g g-Whitespace">   </span><span class="mi">3811</span>         <span class="k">raise</span> <span class="n">InvalidIndexError</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
<span class="ne">-&gt; </span><span class="mi">3812</span>     <span class="k">raise</span> <span class="ne">KeyError</span><span class="p">(</span><span class="n">key</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">err</span>
<span class="g g-Whitespace">   </span><span class="mi">3813</span> <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
<span class="g g-Whitespace">   </span><span class="mi">3814</span>     <span class="c1"># If we have a listlike key, _check_indexing_error will raise</span>
<span class="g g-Whitespace">   </span><span class="mi">3815</span>     <span class="c1">#  InvalidIndexError. Otherwise we fall through and re-raise</span>
<span class="g g-Whitespace">   </span><span class="mi">3816</span>     <span class="c1">#  the TypeError.</span>
<span class="g g-Whitespace">   </span><span class="mi">3817</span>     <span class="bp">self</span><span class="o">.</span><span class="n">_check_indexing_error</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>

<span class="ne">KeyError</span>: 19
</pre></div>
</div>
</div>
</div>
</section>
<section id="other-resource">
<h2>Other Resource<a class="headerlink" href="#other-resource" title="Link to this heading"></a></h2>
<p>NCBI:</p>
<ul class="simple">
<li><p>Explore Virus Data <a class="footnote-reference brackets" href="#id28" id="id14" role="doc-noteref"><span class="fn-bracket">[</span>13<span class="fn-bracket">]</span></a></p></li>
<li><p>NCBI Visual Data Dashboard <a class="footnote-reference brackets" href="#id29" id="id15" role="doc-noteref"><span class="fn-bracket">[</span>14<span class="fn-bracket">]</span></a></p></li>
</ul>
</section>
<section id="reference">
<h2>Reference<a class="headerlink" href="#reference" title="Link to this heading"></a></h2>
</section>
</section>
<hr class="footnotes docutils" />
<aside class="footnote-list brackets">
<aside class="footnote brackets" id="id16" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id1">1</a><span class="fn-bracket">]</span></span>
<p>ICTV. (2024). About Virus Taxonomic Classification. (<a class="reference external" href="https://ictv.global/taxonomy/about">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id17" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id2">2</a><span class="fn-bracket">]</span></span>
<p>Elliot J Lefkowitz, Donald M Dempsey, Robert Curtis Hendrickson, Richard J Orton, Stuart G Siddell, Donald B Smith, Virus taxonomy: the database of the International Committee on Taxonomy of Viruses (ICTV), Nucleic Acids Research, Volume 46, Issue D1, 4 January 2018, Pages D708–D717, (<a class="reference external" href="https://doi.org/10.1093/nar/gkx932">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id18" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id3">3</a><span class="fn-bracket">]</span></span>
<p>wiki. (year). Virus classification. (<a class="reference external" href="https://en.wikipedia.org/wiki/Virus_classification">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id19" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span>4<span class="fn-bracket">]</span></span>
<span class="backrefs">(<a role="doc-backlink" href="#id4">1</a>,<a role="doc-backlink" href="#id6">2</a>)</span>
<p>ICTV. (2024). Master Species Lists (MSL). (<a class="reference external" href="https://ictv.global/msl">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id20" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id5">5</a><span class="fn-bracket">]</span></span>
<p>ICTV. (2024). Virus Metadata Resource (VMR). (<a class="reference external" href="https://ictv.global/vmr">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id21" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id7">6</a><span class="fn-bracket">]</span></span>
<p>ICTV. (2024). Current ICTV Taxonomy Release. (<a class="reference external" href="https://ictv.global/taxonomy">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id22" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id8">7</a><span class="fn-bracket">]</span></span>
<p>ICTV. (2024). Visual Taxonomy Browser. (<a class="reference external" href="https://ictv.global/taxonomy/visual-browser">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id23" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id9">8</a><span class="fn-bracket">]</span></span>
<p>Sykes J. E. (2014). Feline Immunodeficiency Virus Infection. Canine and Feline Infectious Diseases, 209–223. (<a class="reference external" href="https://doi.org/10.1016/B978-1-4377-0795-3.00021-1">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id24" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id10">9</a><span class="fn-bracket">]</span></span>
<p>ICTV. (2023). History of the taxon. (<a class="reference external" href="https://ictv.global/taxonomy/taxondetails?taxnode_id=202305029">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id25" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id11">10</a><span class="fn-bracket">]</span></span>
<p>ICTV. (year). The ICTV Report on Virus Classification and Taxon Nomenclature. (<a class="reference external" href="https://ictv.global/report/chapter/retroviridae/retroviridae/lentivirus">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id26" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id12">11</a><span class="fn-bracket">]</span></span>
<p>Siddell, S. G., Smith, D. B., Adriaenssens, E., Alfenas-Zerbini, P., Dutilh, B. E., Garcia, M. L., Junglen, S., Krupovic, M., Kuhn, J. H., Lambert, A. J., Lefkowitz, E. J., Łobocka, M., Mushegian, A. R., Oksanen, H. M., Robertson, D. L., Rubino, L., Sabanadzovic, S., Simmonds, P., Suzuki, N., Van Doorslaer, K., … Zerbini, F. M. (2023). Virus taxonomy and the role of the International Committee on Taxonomy of Viruses (ICTV). The Journal of general virology, 104(5), 001840. (<a class="reference external" href="https://doi.org/10.1099/jgv.0.001840">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id27" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id13">12</a><span class="fn-bracket">]</span></span>
<p>Kuhn, J. H., Bao, Y., Bavari, S., Becker, S., Bradfute, S., Brister, J. R., Bukreyev, A. A., Chandran, K., Davey, R. A., Dolnik, O., Dye, J. M., Enterlein, S., Hensley, L. E., Honko, A. N., Jahrling, P. B., Johnson, K. M., Kobinger, G., Leroy, E. M., Lever, M. S., Mühlberger, E., … Nichol, S. T. (2013). Virus nomenclature below the species level: a standardized nomenclature for natural variants of viruses assigned to the family Filoviridae. Archives of virology, 158(1), 301–311. (<a class="reference external" href="https://doi.org/10.1007/s00705-012-1454-0">web page</a>)</p>
</aside>
<aside class="footnote brackets" id="id28" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id14">13</a><span class="fn-bracket">]</span></span>
<p>NCBI. (2024). Explore Virus Data: Feline immunodeficiency virus. (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&amp;amp;VirusLineage_ss=Feline%20immunodeficiency%20virus,%20taxid:11673">web pate</a>)</p>
</aside>
<aside class="footnote brackets" id="id29" role="doc-footnote">
<span class="label"><span class="fn-bracket">[</span><a role="doc-backlink" href="#id15">14</a><span class="fn-bracket">]</span></span>
<p>NCBI. (2024). NCBI Visual Data Dashboard. (<a class="reference external" href="https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/">web page</a>)</p>
</aside>
</aside>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="statlab_corr_spearman_rho.html" class="btn btn-neutral float-left" title="&lt;no title&gt;" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="changelog.html" class="btn btn-neutral float-right" title="Changelog" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Y. Hsu.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>